### ISSUE TITLE
kobo aura hd Chinese mobi gibberish

### ISSUE THREAD

**Issue by shuqian** [koreader/koreader#5599]
* KOReader version:v2019.11 (in fact all of v2019 )
* Device:kobo aura hd

#### Issue
most of my mobi books gibberish (utf_8) simple chinese font Microsoft YaHei
![](https://ftp.bmp.ovh/imgs/2019/11/70aed040877e3cf7.jpeg)
![](https://ftp.bmp.ovh/imgs/2019/11/05559114952aaa76.jpeg)

**Comment by Frenzie** [comment]
Are any of these available to check? What if they're converted to EPUB?

**Comment by poire-z** [comment]
Mobi support in crengine is fully in the following file - and it looks like there's some `encoding` field in some header that states if content is cp1252 vs UTF8:
https://github.com/koreader/crengine/blob/master/crengine/src/pdbfmt.cpp#L127
Looks like this encoding field is not used anywhere.

**Comment by shuqian** [comment]
@Frenzie converted to EPUB it is Ok ,not gibberish

**Comment by poire-z** [comment]
> Looks like this encoding field is not used anywhere.

But there's code that check the content, if it is valid UTF-8, and if not, detect the encoding from the bytes, which might end up being detected as russian :)

We'd need some sample of a problematic .mobi to know more about what's happening.

**Comment by shuqian** [comment]
@poire-z Hi,I give u two mobi files [https://send.firefox.com/download/6fac62532826ad6b/#PaK0CKze60r7vlrevjW5Xw](url)  
file angularXXXX.mobi not gibberish,but file dockerXXXXX.mobi is gibberish in Chinese

**Comment by poire-z** [comment]
OK, so it's the utf8 detection that detects the first 16K of the content is not fully valid UTF8.
There are 5 or 6 invalid UTF8 bytes among them, and that's enough to not parse as UTF8.
Extracted from the HTML, shown by Firefox:
<kbd>![image](https://user-images.githubusercontent.com/24273478/69158849-52237500-0ae7-11ea-8e3f-761df440968d.png)</kbd>
Shown by vim as utf8 on a latin1 terminal:
<kbd>![image](https://user-images.githubusercontent.com/24273478/69158872-5d76a080-0ae7-11ea-9806-675337141918.png)</kbd>
xxd:
```
00000000: 342e 322e 32e3 8010 e4b9 a680 e995 9ce5  4.2.2...........
00000010: 838f e5b1 823c 2f61 3e0a 0ae7 acac 2031  .....</a>..... 1
00000020: 3220 e7ab a0e3 8080 e7bc 96e6 8e92 e380  2 ..............
00000030: 81e9 9b86 e7be a4e5 928c e7ae a1e7 3e3c  ..............><
00000040: 7020 7769 6474 6890 863c 2f61 3e3c 2f70  p width..</a></p
00000050: 3e0a                                     >.
```
Your angular mobi does not show any such problem.

Can you upload your Docker.mobi converted as EPUB - just to confirm these invalid bytes are also in the clear text HTML that you converter will have added into the EPUB - and that's the original bad content fault - and not some uncompression/decoding error?

If we simulate a successful utf8 detection, we get this:
<kbd>![image](https://user-images.githubusercontent.com/24273478/69160017-1be6f500-0ae9-11ea-9a1a-1258c070b0de.png)</kbd><kbd>![image](https://user-images.githubusercontent.com/24273478/69160098-42a52b80-0ae9-11ea-9689-b980d3594873.png)</kbd>

If the Docker.mobi loads and shows fine in another reader device or software, please post a screenshot/photo of the Table of Content at start of book, showing the entries `4.2.2` and `12.`, to see what they make out of these invalid bytes there.

So, dunno, may be other readers would show them well because they are more relaxed in the UTF8 detection.

**Comment by Frenzie** [comment]
@poire-z Is Mobi also HTML? If so, is there perhaps a `<meta charset="UTF-8">` or some such which might override automatic detection?

**Comment by poire-z** [comment]
It contains a (single file I think) HTML, inside a binary/compressed package.

As mentionned above, there might be a flag that mentions the encoding - but crengine does not use it and goes thru some autodetection:
https://github.com/koreader/crengine/blob/cb3b2da13b70d8bea86ef70bf959703b02f58962/crengine/src/pdbfmt.cpp#L1199-L1214
which is at:
https://github.com/koreader/crengine/blob/cb3b2da13b70d8bea86ef70bf959703b02f58962/crengine/src/crtxtenc.cpp#L1901-L1931

Dunno much more than this about PDB/MOBI :)

**Comment by shuqian** [comment]
@poire-z @Frenzie 
this is the file i converted as EPUB (Calibre) https://send.firefox.com/download/95bc271e53eb540a/#_bIMBs9AHi5ynT2qy31uIQ

![](https://ftp.bmp.ovh/imgs/2019/11/24494dbe8abbd7e0.png)
Pinyin input method ：jing(镜) xiang(像) ceng(层)   管理(li)

the mobi file inside we can rename Docker.mobi to Docker.rar then extract it,you will see it

![](https://ftp.bmp.ovh/imgs/2019/11/a1acc29c64b018a5.png)

Thank u

**Comment by Frenzie** [comment]
@poire-z I think that explains why other programs are doing better: they just apply that attribute instead of doing autodetection (in this case).

**Comment by NiLuJe** [comment]
Warning: The meta in the ePub conversion is probably Calibre honoring the PDB/Mobi metadata in a way that makes sense for ePub.

**Comment by NiLuJe** [comment]
c.f., codec/codepage in https://github.com/kevinhendricks/KindleUnpack/blob/master/lib/mobi_header.py

**Comment by poire-z** [comment]
Wow, that .mobi opens with 7zip, and shows something that looks like an EPUB container (except not fully). It has a epub.html, which looks sane:
<kbd>![image](https://user-images.githubusercontent.com/24273478/69169275-5dcb6780-0af8-11ea-8ba0-2445e8756214.png)</kbd>
```
0000000: 342e 322e 32e3 8080 e995 9ce5 838f e5b1  4.2.2...........
0000010: 823c 2f61 3e0a 0ae7 acac 2031 3220 e7ab  .</a>..... 12 ..
0000020: a0e3 8080 e7bc 96e6 8e92 e380 81e9 9b86  ................
0000030: e7be a4e5 928c e7ae a1e7 9086 3c2f 613e  ............</a>
0000040: 3c2f 703e 3c70 2063 6c61 7373 3d27 746f  </p><p class='to
0000050: 632d 6c65 7665 6c2d 3227 3e0a 0a         c-level-2'>..
```
So, I don't know, there might be a decoding bug while decompressing/decoding the .mobi if we dont get the correct HTML content.

**Comment by NiLuJe** [comment]
KU says it's indeed flagged as `utf-8`, which, duh :D.

It's also a combination M6/KF8 mobi built by KindleGen (with the original source bundled). This is supposed to harmlessly degrade to M6 on older renderer, but, who knows.

**Comment by NiLuJe** [comment]
The rawml looks sane, I can't see anything obviously wrong with it. Then again, I don't read CJK, even less-so when it's in hex :D.

![zh_m6_rawml](https://user-images.githubusercontent.com/111974/69170639-bd2a7700-0afa-11ea-82a2-8a68ec6a3830.png)

**Comment by NiLuJe** [comment]
TL;DR: Mobi is not great, and CRe's Mobi support even less so.

Use an ePub version. if you don't have any on hand, use KindleUnpack, it'll spit out the source in kindlegensrc.zip ;).

(In addition to rebuilding a sane ePub out of it, even if it didn't happen to bundle the source ;)).

(Which is *maybe* bits of what 7-zip picked up on your end, @poire-z).

**Comment by poire-z** [comment]
Looks like it is:
![image](https://user-images.githubusercontent.com/24273478/69171488-5908b280-0afc-11ea-97a7-d2196617b499.png)
So, there's this readable stuff (whose epub.html might be ok), and some other version of that that I can't access, which might be bad?

**Comment by NiLuJe** [comment]
There's actually three copies of the content ^^.

* The source ePub (which is optional, and usually stripped on delivery to device in an Amazon->Kindle workflow).
* The legacy Mobi7 (M7) markup.
* The KF8 markup.

Wrapped in the usual, supposedly backward compatible Palm DB format.

**Comment by poire-z** [comment]
And it goes thru this while decompression stuff:
https://github.com/koreader/crengine/blob/cb3b2da13b70d8bea86ef70bf959703b02f58962/crengine/src/pdbfmt.cpp#L421-L458

**Comment by NiLuJe** [comment]
If someone were to need some guidance, I consider the canonical implementation to be KindleUnpack ;).

There's also obviously Calibre's, which is derived from KU.

And there's also https://github.com/bfabiszewski/libmobi, which I've never personally used, but @bfabiszewski has done some pretty solid work over the years in Kindle-land ;).

**Comment by poire-z** [comment]
Had an other look, and it's not really the utf8 detection or trusting some flag that is at fault.
It's the decoding of the records in the pdb, the boundaries of the records, that get crap.
![image](https://user-images.githubusercontent.com/24273478/69189018-eb6d7e00-0b1d-11ea-9916-87bd1963385f.png)
![image](https://user-images.githubusercontent.com/24273478/69189045-f9230380-0b1d-11ea-8539-8e92a64dce56.png)
So, this happening on the first 16K might cause the failure of utf8 detection.

But even the other "good" Angular book that does not fail the utf8 detection has that (a little less) crap at start ot end of blocks, so, it's not obvious to me as I don't read chinese, but may be it shows in the content? No chinese char missing?

I tried to follow & compare the code with libmobi... the "lz77" decoding seems ok/similar.
But I got lost in the header/extradataFlags / removeExtraData vs libmobi's read.c mobi_parse_mobiheader/mobi_get_record_extrasize...
The mobi header size seems free style with the many variations.
crengine was doing some extraDataFlags read only [for some specific header sizes](https://github.com/koreader/crengine/blob/cb3b2da13b70d8bea86ef70bf959703b02f58962/crengine/src/pdbfmt.cpp#L193-L195) - while this book gets one larger than that - but just forcing it changes nothing.

So, giving up :|

**Comment by poire-z** [comment]
Uh!
I just had a final try after cleaning all my debug stuff, with just:
```diff
--- a/crengine/src/pdbfmt.cpp
+++ b/crengine/src/pdbfmt.cpp
@@ -190,7 +190,7 @@ struct MobiPreamble : public PalmDocPreamble
             return false; // unsupported type
         if ( mobiEncryption!=0 )
             return false; // encryption is not supported
-        if (hederLength == 0xE4 || hederLength == 0xE8) {
+        if (hederLength >= 0xE4 || hederLength == 0xE8) {
             stream->Seek(242-180, LVSEEK_CUR, NULL);
             stream->Read(&extraDataFlags);
             if ( cnv.lsf() )
```
and that Docker book displays chinese!

**Comment by NiLuJe** [comment]
That makes half of that test dead code, since 0xE8 > 0XE4 ;).

**Comment by poire-z** [comment]
Yes yes :) (but are you sure that's still true in Kindle-land ?!)
@shuqian: here's [libcrengine+cre-mobi-fix.zip for Kobo](https://github.com/koreader/koreader/files/3866366/libcrengine%2Bcre-mobi-fix.zip) containing libcrengine.so and libkoreader-cre.so that you can drop in koreader/libs/ on your Kobo with a recent koreader, so you can check it that indeed fixes your issue with the Docker book, and others. Tell us how it does.

**Comment by NiLuJe** [comment]
That vaguely looks like [this](https://github.com/kevinhendricks/KindleUnpack/blob/743728e541d5bf2dbc8127f4334dc1530eca0a30/lib/mobi_header.py#L834) FWIW, so, eh.

**Comment by shuqian** [comment]
@poire-z I just tried，It is fixed ,not gibberish again.
so great. thank you all.  
;).

**Comment by poire-z** [comment]
Good, thanks. I'll add that fix in a few days.

Btw, @NiLuJe , why in https://github.com/koreader/crengine/commit/3433c077f4b07e50e5d06232d6c7cd28ea53a2b0#diff-177cf4caa03bed6297a8bd322d30fdd3 could you use `PDBHdr hdr = { 0 };` `EReaderHeader preamble = { 0 };`, but not `MobiPreamble preamble = { 0 };` (the compiler complains), so you had to use just `MobiPreamble preamble = {};` ?

**Comment by NiLuJe** [comment]
There's an union in the second, or something like that, IIRC.

**Comment by NiLuJe** [comment]
(Re-opening until the fix actually lands ;))
